Add totals to Courses report

donnapep opened the pull request:
Partial fix for #4834.

### Changes proposed in this Pull Request
Adds totals to the _Course_, _Completed_ and _Average Grade_ column headers of the Courses report.

### Testing instructions
* Add several courses with lessons that have quizzes.
* Take and grade the quizzes.
* Complete the courses.
* For the Courses report, check that the _Course_, _Completed_ and _Average Grade_ column headers display the correct totals. (Note that 0% entries for _Average Grade_ can be ignored as they are not included in the total average.)
* Trash and/or change the status to _Draft_ for one or more courses.
* Ensure that the _Course_ column header total is updated and correct.
* In _Student Management_, reset the progress of one or more students who completed a course.
* Ensure that the _Completed_ and _Average Grade_ column header totals are updated and correct.

LavaGolem commented on includes/class-sensei-analysis-overview-list-table.php:
```diff
@@ -56,7 +56,11 @@ function get_columns() {
+					'average_percent' => sprintf(
+						// translators: Placeholder value is the average grade of all courses.
+						__( 'Average Grade (%d%%)', 'sensei-lms' ),
+						esc_html( Sensei()->grading->get_courses_average_grade() )
```
@donnapep  Hi Donna I know this is just draft, but wanted to point out something that I've discovered when doing totals for Students reports: 
```get_columns()``` function is being called 3 times. And only once to generate the table, other two times if from some VIP functions. 
The problem with this is that  get_courses_average_grade() function is also called 3 times. I tested it locally by print_r in the ```get_courses_average_grade()``` function, so you can see it here. On single page load it is called 3 times:  (Very savage but quick approach :D )
<img width="446" alt="image" src="https://user-images.githubusercontent.com/7208249/157429656-ba780555-80bd-40d4-ae76-05a67feeafac.png">

I solved that by moving the call to Sensei()->grading->get_courses_average_grade() in prepare data function. [Here](https://github.com/Automattic/sensei/pull/4865/files#diff-6ce1dc075c4832d5bf050e621bcfecbab91d97504e8bc565f4f476b1ea2f654dR214)

and storing the data to class properties:[Here]( https://github.com/Automattic/sensei/pull/4865/files#diff-6ce1dc075c4832d5bf050e621bcfecbab91d97504e8bc565f4f476b1ea2f654dR31
)

And using them like [this ](https://github.com/Automattic/sensei/pull/4865/files#diff-6ce1dc075c4832d5bf050e621bcfecbab91d97504e8bc565f4f476b1ea2f654dR102)

donnapep replied:
@LavaGolem Thanks! I went down a rabbit hole trying to figure out why it's calling it multiple times. I *think* it may be related to [this issue](https://core.trac.wordpress.org/ticket/34564), but the solution was not obvious so I went with an approach similar to your's. The only difference is that I [added the variable to the Grading class](https://github.com/Automattic/sensei/pull/4904/commits/71e357afe814b36bc0d985f156c2436cd9a37a58#diff-15ca8a2e2992e136808c71f8edc63942547a9b2a1e9535f291e4fc27c3fbaf1fR27) instead, which seems like a more appropriate place for it, and then [checked if the variable is set](https://github.com/Automattic/sensei/pull/4904/commits/71e357afe814b36bc0d985f156c2436cd9a37a58#diff-15ca8a2e2992e136808c71f8edc63942547a9b2a1e9535f291e4fc27c3fbaf1fR1221) before querying the data again.

WDYT?

m1r0 replied:
I think instead of using a variable in the Grading class, it would be better to fix the `get_columns` method so it's not called for each column. Here's a patch:

```patch
diff --git a/includes/class-sensei-analysis-overview-list-table.php b/includes/class-sensei-analysis-overview-list-table.php
index 98b63d828..18843bc38 100755
--- a/includes/class-sensei-analysis-overview-list-table.php
+++ b/includes/class-sensei-analysis-overview-list-table.php
@@ -50,6 +50,10 @@ class Sensei_Analysis_Overview_List_Table extends Sensei_List_Table {
         */
        function get_columns() {

+               if ( $this->columns ) {
+                       return $this->columns;
+               }
+
                switch ( $this->type ) {
                        case 'courses':
                                $columns = array(
@@ -89,7 +93,10 @@ class Sensei_Analysis_Overview_List_Table extends Sensei_List_Table {
                // Backwards compatible filter name, moving forward should have single filter name
                $columns = apply_filters( 'sensei_analysis_overview_' . $this->type . '_columns', $columns, $this );
                $columns = apply_filters( 'sensei_analysis_overview_columns', $columns, $this );
-               return $columns;
+
+               $this->columns = $columns;
+
+               return $this->columns;
        }

        /**
```

donnapep replied:
Oh, you so smart! 🕺🏻 https://github.com/Automattic/sensei/pull/4904/commits/bfc499294d6f5614a33e7e5ffa7633bc43be6f22

m1r0 commented:
I'm not sure if this is an issue with the totals or an issue with the rows not showing the correct data, but I've noticed the following:

<img width="1299" alt="Screenshot on 2022-03-11 at 02-39-10" src="https://user-images.githubusercontent.com/1612178/157779474-6d96f503-feb9-4c94-81b9-57bdc599a477.png">

Here the average grade is 66% which doesn't look right. I have one row of 50% and one row of 100%, so it should be 75%, correct? I've tracked down the issue to the "Course with modules". That course has a user that is enrolled but didn't complete any lessons (percent = 0).

<img width="1287" alt="Screenshot on 2022-03-11 at 02-41-34 (1)" src="https://user-images.githubusercontent.com/1612178/157779750-0e4f3aeb-8fb0-4b03-bb73-a26cea27c8fd.png">

If I unenroll and reset that user from that course, the average goes to 75%.

donnapep commented:
@m1r0 Are you still able to reproduce it? Can you give me a database dump? I think it will be faster that way than me trying to reproduce it.

To be honest, I'm not 100% confident that any of the tables are reporting _Average Grade_ correctly. 😅  As I understand it, this column may show a 0 in the following cases:

1. The student legitimately received a grade of 0 in a course that had quizzes.
2. The student is still taking the course but hasn't completed it yet.
3. The student completed the course but the course had no quizzes. Should they really have a grade in that case?

Regardless of whether the values are right or wrong, I'm using a similar query to what is being used to fetch the row data. So if it's wrong, it should at least be consistently wrong. 😄

m1r0 commented:
I think we are using the wrong meta here. The `percent` meta represents the course completion percent (how many lessons are completed for that course) and not the grading percent.

You could test this by just completing a lesson with no quiz and see the percentage increase. I think [this code](https://github.com/Automattic/sensei/blob/master/includes/class-sensei-analysis-course-list-table.php#L462-L474) might help.

> it should at least be consistently wrong. 😄

🤣

donnapep commented:
> I think we are using the wrong meta here. The percent meta represents the course completion percent (how many lessons are completed for that course) and not the grading percent.

Ugh, you're right.

I initially fixed this [here](https://github.com/Automattic/sensei/pull/4904/commits/79d82d30389d6ce0644c7d6e1e2620a262f9cfdf), but still found that I had some discrepancies in my numbers that were a result of the following data integrity issues:
- Some of my lessons had a `_lesson_course` value that contained a course ID that did not exist. So I added a check for this in my query.
- I also found I had grade data for some lessons but `_quiz_has_questions` meta didn't exist for those lessons. I'm guessing I removed the quiz questions at some point and Sensei didn't clean up the grade data. 😩  I needed to add a check for this as well.

When you're testing this, note that you won't be able to drill down into each lesson and inspect the average grade data, because it's not correct either. 😮 But I'm now displaying "N/A" in the average grade column for all courses that legitimately don't have any grade data, so you should be able to sum up all the values in _Average Grade_ and divide by the number of courses with average grades

I'll update tests and resolve conflicts next week.

m1r0 reviewed (approved):
I did some testing and it works well. 👍 I've noticed the `days to completion` total is not implemented. Are we dropping it or it will be added in a future PR?

donnapep commented:
> I've noticed the `days to completion` total is not implemented

@m1r0 That column wasn't there until I resolved conflicts, so I'll do it in a separate PR so the other column total PRs can pull in my changes. BTW, I saw the comment about using subqueries in another PR. I'll admit that I didn't test with tons of data, so feel free to optimize this query if it's slow. My tests should catch any potential breakage if we refactor the query.
